### DISCUSSION TITLE
white globe and zillions of openGL warning

### DISCUSSION THREAD

**Discussion by VirtualSim** [pelicanmapping/osgearth#2398] [Support Q&A] [answered]
Hi there,

I have OSG 3.6.5, x64 (under vs2017) with ENABLE_GL_3 ON (GL_1 & GL_2 OFF)
I have build osgEARTH 3.5 in x64 (I do not have geocoder, since it cannot be compiled, many errors)

What I have, when I run any sample: A while globe !
And the console shows opengl warnings like crazy.

So, where is the problem? OSG build ? osgEarth build ? missing or faulty dll ?

Thanks.

Warning: detected OpenGL error 'invalid operation' at start of State::apply()
Warning: detected OpenGL error 'invalid operation' at start of State::apply(StateSet*)
Warning: TexEnv::apply(State&) - not supported.
Warning: TexEnv::apply(State&) - not supported.
Warning: TexEnv::apply(State&) - not supported.
Warning: detected OpenGL error 'invalid operation' at start of State::apply()
Warning: detected OpenGL error 'invalid operation' at start of State::apply(StateSet*)
Warning: TexEnv::apply(State&) - not supported.
Warning: TexEnv::apply(State&) - not supported.
Warning: TexEnv::apply(State&) - not supported.

**Comment by gwaldron** [comment]
Please post the output of `osgearth_version --caps`. Thanks.
You can also run one of the samples with `--gldebug` for a little more information.

**Reply by DongShuo-96** [reply]
I also have a similar problem, which I believe is caused by OSG and theoretically has little to do with OSG
Problem: If I use osg3.7 (only gl3) and osgearth3.4, the scene runs normally and the font display is also normal. However, if I use osg3.6.5 (gl3+gl2) and osgearth3.4, the scene runs normally but the font display crashes. All the fonts in the state information are lost, which will cause the program to crash abnormally. The program terminal keeps outputting "Warning: TexEnv:: apply (State&) - not supported", The example run of OSG also shows this phenomenon, and the entire Earth scene is quite sluggish. I need OSG (gl2) because I am compiling OSGOCEAN.
Looking forward to your reply on this issue.

**Comment by VirtualSim** [comment]
Thanks.
I found that on the first line:  Warning, OpenSceneGraph does not define OSG_GL3_AVAILABLE
The osg 3.6.5 I tried to build with USE_OPENGL_3 ON produced the continuous opengl warnings. The binary install I have found seems not to have GL3 ON. Si, back to square 1.
Do you know why OSG with GL3 switched to ON produces that many warning ?
Anyway, with or without warning, globe is still white.
I just downloaded the osgEarth from GitHub (3.5.0)
--gldebug does not produces more output:
[osgEarth]* [Capabilities] Warning, OpenSceneGraph does not define OSG_GL3_AVAILABLE; the application may not function properly
[osgEarth]* [EarthSerializer2] class osgEarth::TMSImageLayer "ReadyMap 15m Imagery" : Resource unavailable: Failed to read configuration from http://readymap.org/readymap/tiles/1.0.0/7/
[osgEarth]* [EarthSerializer2] class osgEarth::TMSElevationLayer "ReadyMap 90m Elevation" : Resource unavailable: Failed to read configuration from http://readymap.org/readymap/tiles/1.0.0/116/
[osgEarth]* [SimpleSkyNode] Failed to load moon texture from moon_1024x512.jpg

D:\OpenSceneGraph-3.6.5\osgearth-3.5\vc14_x64\bin\Release\osgearth_eci.exe (process 3600) exited with code 0.
To automatically close the console when debugging stops, enable Tools->Options->Debugging->Automatically close the console when debugging stops.

**Reply by gwaldron** [reply]
I suspect something went wrong with the way you build OSG. (I am not familiar with a "USE_OPENGL_3" setting...where do you see this?)

If you are building OSG from scratch and want to build in GL3 mode, you just need to set the OPENGL_PROFILE cmake flag to "GL3" and then cmake will set all the other vars for you.

In the master branch of osgEarth we removed the GL3 warning and are trying to make it easier to support either a GL2 or GL3 build for most functions in osgEarth. You can try that if you wish.

The warning messages come from OSG and you see then b/c OSG is trying to execute OpenGL fixed-function pipeline (FFP) calls when they are not available. Are you loading a model?

**Comment by VirtualSim** [comment]
Thanks again.
I have rebuilt OSG from scratch in x64 and GL2 only. Demos as running fine without any warning.
I have also downloaded osgEarth 3.3 which is not using GL3. Rebuilt it with several dependencies + shaders without any error.
But at start, it displays the following and crashes:
[osgEarth]* [EarthSerializer2] class osgEarth::TMSImageLayer "ReadyMap 15m Imagery" : Resource unavailable : Failed to read configuration from http://readymap.org/readymap/tiles/1.0.0/7/
[osgEarth]* [EarthSerializer2] class osgEarth::TMSElevationLayer "ReadyMap 90m Elevation" : Resource unavailable : Failed to read configuration from http://readymap.org/readymap/tiles/1.0.0/116/
[osgEarth]* [SimpleSkyNode] Failed to load moon texture from moon_1024x512.jpg

If I open the earth file with SXEarth software, it also display a white globe but no problem having a bing/google map with default SXEarth software.
My question: is it the shaders the problem or the build ? Can you provide the version of the dependencies needed to display tiled textures ?
Thanks a lot.
![err](https://github.com/gwaldron/osgearth/assets/18327137/c9460df8-7c37-478d-9dff-424092927ccc)

**Reply by gwaldron** [reply]
This message
```
Resource unavailable : Failed to read configuration from http://readymap.org/readymap/tiles/1.0.0/7/
```
simply means your machine could not access the URL. (Try https, by the way.)
You can try setting the envvar `OSGEARTH_HTTP_DEBUG=1` to see if there's any additional information.

**Comment by VirtualSim** [comment]
Thanks.
I tried the https but without success:

(readymap.earth)...[osgEarth]* [HTTPClient] HTTP debugging enabled
[osgEarth]  [HTTPClient] GET(0, ) : "https://readymap.org/readymap/tiles/1.0.0/116/" (Thu, 01 Jan 1970 00:00:00 GMT) t=9.55e-05s
[osgEarth]  [TerrainLayer] Layer "readymap_elevation" Failed to create "tms" driver, but a cache may exist, so falling back on cache-only mode.
[osgEarth]* [EarthSerializer2] Layer "readymap_elevation" : Error : Failed to open a cache for layer because cache_only policy is in effect and bin [43e1f904] could not be located.
[osgEarth]  [HTTPClient] GET(0, ) : "https://readymap.org/readymap/tiles/1.0.0/7/" (Thu, 01 Jan 1970 00:00:00 GMT) t=3.15e-05s
[osgEarth]  [TerrainLayer] Layer "readymap_imagery" Failed to create "tms" driver, but a cache may exist, so falling back on cache-only mode.
[osgEarth]* [EarthSerializer2] Layer "readymap_imagery" : Error : Failed to open a cache for layer because cache_only policy is in effect and bin [a4c99ecd] could not be located.

Any other clue ?
Thanks a lot.

**Reply by gwaldron** [reply]
It is just failing to load the URL. Is your CURL built with OpenSSL support?

**Comment by VirtualSim** [comment, marked as answer]
THANKS !!!
It was that. I rebuild Curl with OpenSSL + LIBSSH (for now, osgEarth 2.8 under OSG 3.4) and I got no more problem.
I will do the same with osgEarth 3.5 with OSG 3.6 with GL3 in x64 and see if I can finally see something.
I guess all this comes from the fact that https is now used. 
I hope this will be helpful for others as I am not sure all precompiled curl libs include OpenSSL.
Thanks again, you saved my week :-)

**Reply by gwaldron** [reply]
Glad to hear it :)
